[Breaking] Bump minimum TypeScript version to 5.1.0

devjiwonchoi opened the pull request:
This PR bumps the minimum TS version required for Next.js to `5.1.0`, which is the version that supports React Components that return a promise via `JSX.ElementType`.

https://devblogs.microsoft.com/typescript/announcing-typescript-5-1/#decoupled-type-checking-between-jsx-elements-and-jsx-tag-types

devjiwonchoi committed:
[Breaking] Bump minimum TypeScript version to 5.0.0

devjiwonchoi commented:
* **#84384** <a href="https://app.graphite.dev/github/pr/vercel/next.js/84384?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/vercel/next.js/84384?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `canary`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

ijjk commented:
## Tests Passed

<!-- ## Failing test suites -- >

eps1lon commented:
Why 5.0 and not [5.3 that's currently lowest supported by DT](https://github.com/DefinitelyTyped/DefinitelyTyped#support-window)?

Especially 5.1 would be a nice milestone since that's the version that shipped support for returning Promises from React Components.

devjiwonchoi commented:
> Why 5.0 and not [5.3 that's currently lowest supported by DT](https://github.com/DefinitelyTyped/DefinitelyTyped#support-window)?
> 
> Especially 5.1 would be a nice milestone since that's the version that shipped support for returning Promises from React Components.

5.1 sounds good

devjiwonchoi committed:
Change minimum to v5.1.0
As this is the version that supports React Components that return a promise via JSX.ElementType.
x-ref: https://devblogs.microsoft.com/typescript/announcing-typescript-5-1/#decoupled-type-checking-between-jsx-elements-and-jsx-tag-types

Co-authored-by: Sebastian "Sebbie" Silbermann <silbermann.sebastian@gmail.com>
